### ISSUE TITLE
Create lesson release cycle plan

### ISSUE THREAD

**Issue by sstevens2** [carpentries/lesson-development-training#188]
Inspired by much discussion, book club, and intermediate software development training, we think we should try to do formal release cycles - collecting changes in development and releasing the new version at once as opposed to merging all the changes into main directly. @anenadic is also thinking about this for one of her lessons and may have a draft/template plan to contribute.

**Comment by tobyhodges** [comment]
Realistically, I do not think this will be ready by the end of this month. If we have time, we should discuss this during our next Maintainer meeting but, if we are not able to define a concrete plan for this at that call I think we should drop it from the beta milestone. Please speak up if you think it's essential that this is resolved before that beta release is made.

**Comment by tobyhodges** [comment]
My first draft of a plan for how this could work:

- every October: create a 'vYYYY' branch (e.g. v2024) to collect changes for next release
- October-May: create issues, discuss them as usual
    - issues labelled type:bug should be fixed by PRs directly to main
    - all other PRs should be made to the vYYYY branch
- June: identify issues still to be addressed in next release, create milestone for these
- June-August: fix issues in release milestone
- September: merge vYYYY branch into main and release new version to Zenodo
    - This would fit with the yearly rotation of the Lead Maintainer role in September - we could hand over to new lead maintainer after release has been made?

---

Questions that immediately spring to mind from this:

1. How could we make it easy to preview the built lesson site from the development branch (which I was calling 'vYYYY' above)?
2. How would we keep track of what has changed in the release branch, to help prevent duplicate issues & PRs being opened based on current main branch content?
3. (Related to the question above) would yearly releases be frequent enough? Should we aim for two per year?

**Comment by anenadic** [comment]
This is great start @tobyhodges. 

> * This would fit with the yearly rotation of the Lead Maintainer role in September - we could hand over to new lead maintainer after release has been made?

Maybe also add into the official responsibilities for the Lead Maintainer.


> My first draft of a plan for how this could work:
> 
> * every October: create a 'vYYYY' branch (e.g. v2024) to collect changes for next release
> * October-May: create issues, discuss them as usual
>   
>   * issues labelled type:bug should be fixed by PRs directly to main
>   * all other PRs should be made to the vYYYY branch
> * June: identify issues still to be addressed in next release, create milestone for these
> * June-August: fix issues in release milestone
> * September: merge vYYYY branch into main and release new version to Zenodo
>   
>   * This would fit with the yearly rotation of the Lead Maintainer role in September - we could hand over to new lead maintainer after release has been made?

> 
> Questions that immediately spring to mind from this:
> 
> 1. How could we make it easy to preview the built lesson site from the development branch (which I was calling 'vYYYY' above)?
This is the million dollar question. I suppose we may need/want to be able to render at least the latest official release that is pointed to by Zenodo, e.g. "v2023" and the next one in preparation "v2024" to preview the latest changes? Or do we? 

> 2. How would we keep track of what has changed in the release branch, to help prevent duplicate issues & PRs being opened based on current main branch content?
If I understand the question correctly, I think we would have to merge "main" with bug fixes back into the next release branch e.g. "v2024" and do this very often so they are aligned as much as possible and the next release branch is (almost) always ahead of the main. They would diverge a bit when bug fixes are applied to "main" but that should then almost immediately be merged into the the next release branch to sync them up. 

> 3. (Related to the question above) would yearly releases be frequent enough? Should we aim for two per year?
While we should probably aim to release often - not sure in terms of effort if we would be able to do it. If it is just a matter of tagging and pushing a button to release (with whatever changes are currently in the release branch), it may be feasible to do it more often. But we may need to do a sprint to finish off issues for the release, and maybe think of the release notes and log - which is extra work.

Some other questions that come to mind:
4. How do we decide what it a bug fix and what is an improvement that can wait? Sometimes it is clear (e.g. a screenshot that is out of date or a typo are bugs) and sometimes (at least) I may tend to wish to merge all fixes onto main without waiting for the official release. 
5. What version naming format do we use? The next two releases would presumably be betas (i.e. "beta-v2023" and "beta-v2024"), while we are still refining the lesson. And do we then we switch to, e.g. "v2025", when the lesson becomes stable. And if we do two releases next year, would they be "beta-v2024" and ""beta-v2024-2".

**Comment by sstevens2** [comment]
> > 2. How would we keep track of what has changed in the release branch, to help prevent duplicate issues & PRs being opened based on current main branch content?

> @anenadic If I understand the question correctly, I think we would have to merge "main" with bug fixes back into the next release branch e.g. "v2024" and do this very often so they are aligned as much as possible and the next release branch is (almost) always ahead of the main. They would diverge a bit when bug fixes are applied to "main" but that should then almost immediately be merged into the the next release branch to sync them up.

I think I interpreted this question differently from Aleks.  I'd suggest we keep track of changes that are in progress in the current release so that we have a human readable list (maybe this is just the milestones?, including completed issues) we can refer to if we get duplicate issues.
Aleks seems to be describing how to make sure the release branch keeps up with main branch bug fixes.  I think that should be us rebasing the release branch based on main every so often (maybe whenever someone works on a PR?) to make sure those changes are included in the release branch.

**Comment by sstevens2** [comment]
No suggestions from me about question 1 about previewing. 

Re number 3, I did wonder if we should try for 2 a year but I think 1 will be more maintainable for us. We could start with one and move up to do if we deem we need more and have the human effort to do that.

> 4. How do we decide what it a bug fix and what is an improvement that can wait? Sometimes it is clear (e.g. a screenshot that is out of date or a typo are bugs) and sometimes (at least) I may tend to wish to merge all fixes onto main without waiting for the official release.

Think this might be up to the maintainer but they could ask for advice around issues they are unsure about.  We could also have the option for any maintainer to push up an issue up to be address immediately instead of waiting for a release.
 
> 6. What version naming format do we use? The next two releases would presumably be betas (i.e. "beta-v2023" and "beta-v2024"), while we are still refining the lesson. And do we then we switch to, e.g. "v2025", when the lesson becomes stable. And if we do two releases next year, would they be "beta-v2024" and ""beta-v2024-2".

Is there a reason to include beta and not just go with 2023 for this year's release (even though we are calling it beta)?  Seems like it would add inconsistent naming but I'm not sure what the advantage would be.

**Comment by anenadic** [comment]
> Is there a reason to include beta and not just go with "v2023" for this year's release (even though we are calling it beta)? Seems like it would add inconsistent naming but I'm not sure what the advantage would be.

I was thinking of that too (as we have the lesson status in the parameter in config.yaml), but the practice in software release is to have "alpha/beta/etc." in the release name/tag.

**Comment by anenadic** [comment]
> > > 2. How would we keep track of what has changed in the release branch, to help prevent duplicate issues & PRs being opened based on current main branch content?
> 
> > @anenadic If I understand the question correctly, I think we would have to merge "main" with bug fixes back into the next release branch e.g. "v2024" and do this very often so they are aligned as much as possible and the next release branch is (almost) always ahead of the main. They would diverge a bit when bug fixes are applied to "main" but that should then almost immediately be merged into the the next release branch to sync them up.
> 
> I think I interpreted this question differently from Aleks. I'd suggest we keep track of changes that are in progress in the current release so that we have a human readable list (maybe this is just the milestones?, including completed issues) we can refer to if we get duplicate issues. Aleks seems to be describing how to make sure the release branch keeps up with main branch bug fixes. I think that should be us rebasing the release branch based on main every so often (maybe whenever someone works on a PR?) to make sure those changes are included in the release branch.

Yes, I think we were answering different part of the question - I'd tag the issues for the next release to indicate they will be addressed or are being addressed in the future and keep the the next release branch in line with main after each PR is merged on the main (ideally).

**Comment by tobyhodges** [comment]
Notes from a conversation with @froggleston:

> How would we keep track of what has changed in the release branch, to help prevent duplicate issues & PRs being opened based on current main branch content?
 
This is always a problem with software release cycles. the good thing is that people tend to open an issue before opening a PR, so you can catch this potential duplication of effort early.
 
> (Related to the question above) would yearly releases be frequent enough? Should we aim for two per year?
 
Twelve months is a long time between releases unless you expect the curriculum to change fairly infrequently. (Toby: after the first stable release next year, I expect/hope that major changes to the curriculum will indeed become less frequent.)

**Comment by tobyhodges** [comment]
> > 6. What version naming format do we use? The next two releases would presumably be betas (i.e. "beta-v2023" and "beta-v2024"), while we are still refining the lesson. And do we then we switch to, e.g. "v2025", when the lesson becomes stable. And if we do two releases next year, would they be "beta-v2024" and ""beta-v2024-2".
> 
> Is there a reason to include beta and not just go with "v2023" for this year's release (even though we are calling it beta)? Seems like it would add inconsistent naming but I'm not sure what the advantage would be.

I had imagined that we would call the upcoming release 'beta' and then adopt version "numbering" from the first stable release (next year). I don't feel strongly about the naming convention, and happy to adopt semantic versioning if you prefer. (I guess the beta version would be 0.2.0, on the assumption that the first alpha version was 0.1.0)

**Comment by tobyhodges** [comment]
> I suppose we may need/want to be able to render at least the latest official release that is pointed to by Zenodo, e.g. "v2023" and the next one in preparation "v2024" to preview the latest changes? Or do we?

I'm not sure we need to keep serving the most-recently-released version: I think the version built from `main` (i.e. that latest release + any bug fixes that have been merged since) is fine: people can build a local version from the files on Zenodo if they really want exactly the version that was released. 

But I certainly think it would be helpful to be able to view the "development" version of the site online, as opposed to only via a local build.

**Comment by anenadic** [comment]
Possibly also related to #486. @sstevens2 and me feel that we have our own release cycle and can close this issue.
